### ISSUE TITLE
[Question] Is there a way to get the session ID?  Is there such a thing in Playwright?

### ISSUE THREAD

**Issue by rlundy** [microsoft/playwright-java#601]
Does Playwright have the concept of a unique session ID for each browser session (or each page, context, whatever)?

If so, is there a way to get the session ID of the current browser/page/context?

**Comment by yury-s** [comment]
No, there is no such think in the public API. Why do you need that?

**Comment by rlundy** [comment]
Hi, I apologize for my tardiness in responding.

I need this because I use Selenium Box (by Element34) for browser sessions.  After a test is run, I use the session ID to mark the session in Selenium Box as passed or failed, so it shows up that way in the Selenium Box API.

It also makes it easier to find the video of the session again for review, because the video is tagged with the session GUID.

It seems like there must be a session ID, since Selenium Box assigns one (a GUID).  Is there a reason it's not exposed in the public API?

**Comment by mxschmitt** [comment]
> It also makes it easier to find the video of the session again for review, because the video is tagged with the session GUID.

The video's file name is a [random GUID](https://github.com/microsoft/playwright/blob/2ec82b9a5cc57fdb68af270897aa4cae8c9db450/src/server/chromium/crPage.ts#L440). If you want to have a deterministic name, you can use [Video.saveAs()](https://playwright.dev/java/docs/api/class-video#video-save-as)

> It seems like there must be a session ID

I would recommend to ask them how they obtain it, probably also random.

**Comment by rlundy** [comment]
OK, I will see what they say.

**Comment by yury-s** [comment]
> I need this because I use Selenium Box (by Element34) for browser sessions. After a test is run, I use the session ID to mark the session in Selenium Box as passed or failed, so it shows up that way in the Selenium Box API.

The sessions is a concept from Selenium Box model and it'd be logical if they provided such an API. There is no such thing as session in the Playwright client api today and you can always compare Browser instances for equality if you need to differentiate between different browsers on the client end. Also the client can pass extra information via headers in `BrowserType.connect`, would that work? Could you provide more details about the scenario that you have or maybe point to a code which needs this?

**Comment by rlundy** [comment]
> The sessions is a concept from Selenium Box model

The session ID is a concept in Selenium, e.g.:
https://www.selenium.dev/selenium/docs/api/dotnet/html/P_OpenQA_Selenium_Remote_RemoteWebDriver_SessionId.htm

And judging from the Java code Element34 sent, it also seems to be a Playwright concept, even if it's not yet a publicly-exposed concept.

The session ID is useful to map a particular browser session to a test, particularly when the session has an associated video recording.  In my case this is helpful because the company I work for creates communication software, so a lot of our tests use two or more browsers to simulate web site visitors and agents who communicate with them.

**Comment by yury-s** [comment]
> And judging from the Java code Element34 sent, it also seems to be a Playwright concept, even if it's not yet a publicly-exposed concept.

In Playwright each client is identified by it connection which created when you connect to the driver and closed when you disconnect.  The code that you cited is misleading in that case we could use `"guid":1` instead of `"guid":"Browser@f5a38d7b-3481-4d30-9208-46d87e232390"` which would be the same for all clients, it works only because we chose to use a guid for that but it may well change.

**Comment by yury-s** [comment]
One important difference between Playwright and WebDriver is that in WebDriver every command is a separate REST request and to let the server know that the requests are addressed to a particular browser instance each command carries `sessionId`. In Playwright all commands flow through the same connection which among other things serves as the identifier but it's lower (transport) level than what Playwright exposes in the API.

**Comment by rlundy** [comment]
So is there a better unique identifier to use in a case like this, where there's a video tied to the browser session?

**Comment by yury-s** [comment]
> where there's a video tied to the browser session?

You can get access to the Video object associated with a Page via [`Page.video()`](https://playwright.dev/java/docs/api/class-page/#page-video) but that's not what you want probably. 

> It seems like there must be a session ID, since Selenium Box assigns one (a GUID). Is there a reason it's not exposed in the public API?

Is there a public documentation describing what kind of session ID they assign to Playwright or source code?

When you connect from you client machine to Selenium Box they create instances of the browsers on your behalf and will control where the files like videos are saved. If the service saves the artifacts somewhere it can then give you back the location, the connection to the service can be used as the session identifier in such case. This is a problem of the service which archives test artifacts and then give you access to them.

**Comment by yangamti** [comment]
Hi @rlundy,

I am facing the same challange, with migration to playwrght using SeleniumBox. Can you please share how you manage to resolve this challange.

My usecase is that I need to link tests executed in SeleniumBox (Which are identified with the session ID) to results that are published to Jira, I do not attached videos into jira ticket instead I add the link to the test video on seleniumBox.
